### ISSUE TITLE
Segmentation fault in PHP 7.0.3 (and earlier versions)

### ISSUE THREAD

**Issue by sisve** [oerdnj/deb.sury.org#246]
#Hi,

I've got some segmentation faults that I am trying to find the root cause for. I am unable to reproduce them from a new blank Ubuntu 14.04 machine and my own compiled version of PHP 7.0.3, while it happens everytime with the packages available from your PPA.

Steps to reproduce:
- add-apt-repository "deb http://ppa.launchpad.net/ondrej/php/ubuntu trusty main"
- apt-key adv --recv-keys --keyserver keyserver.ubuntu.com E5267A6C
- apt-get update
- apt-get install -y php7.0-cli php7.0-dbg php7.0-curl php7.0-gd php7.0-mcrypt php7.0-mysql php7.0-sqlite php7.0-xsl
- php vendor/bin/phpunit

> root@php7dev-ubuntu:~/web# php vendor/bin/phpunit
> PHPUnit 4.8.21 by Sebastian Bergmann and contributors.
> 
> Runtime:	PHP 7.0.3-1+deb.sury.org~trusty+1
> Configuration:	/root/web/phpunit.xml
> 
> .............................................................   61 / 1229 (  4%)
> .............................................................  122 / 1229 (  9%)
> .............................................................  183 / 1229 ( 14%)
> .............................................................  244 / 1229 ( 19%)
> .............................................................  305 / 1229 ( 24%)
> .............................................................  366 / 1229 ( 29%)
> .............................................................  427 / 1229 ( 34%)
> .............................................................  488 / 1229 ( 39%)
> .............................................................  549 / 1229 ( 44%)
> .............................................................  610 / 1229 ( 49%)
> ...............................................Segmentation fault (core dumped)

The backtrace from gdb is available at https://gist.github.com/sisve/0f75be357e557d439c19

I sent an [mail to the internals mailing list][1] to ask how to debug this further; but I am unable to reproduce your compilation result since `php -i` does not include the compilation flags. Are these public somewhere so that I can recompile with more debugging/less optimization?

// Simon

[1]: http://news.php.net/php.internals/91090

**Comment by oerdnj** [comment]
Simon, thanks for looking into this.

The most definitive source of the params are build logs:
https://launchpad.net/~ondrej/+archive/ubuntu/php/+builds?build_text=&build_state=built

The build log for 7.0.3-1 on trusty amd64 can be found here:

https://launchpadlibrarian.net/236379101/buildlog_ubuntu-trusty-amd64.php7.0_7.0.3-1+deb.sury.org~trusty+1_BUILDING.txt.gz

And the configure options for CLI SAPI would be:
```
	mkdir -p cli-build
	cd cli-build
	../configure --build=x86_64-linux-gnu --prefix=/usr --includedir=\${prefix}/include --mandir=\${prefix}/share/man --infodir=\${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --libdir=\${prefix}/lib/x86_64-linux-gnu --libexecdir=\${prefix}/lib/x86_64-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --prefix=/usr --enable-cli --disable-cgi --with-config-file-path=/etc/php/7.0/cli --with-config-file-scan-dir=/etc/php/7.0/cli/conf.d --build=x86_64-linux-gnu --host=x86_64-linux-gnu --libdir=\${prefix}/lib/php --libexecdir=\${prefix}/lib/php --datadir=\${prefix}/share/php/7.0 --program-suffix=7.0 --sysconfdir=/etc --localstatedir=/var --mandir=/usr/share/man --disable-all --disable-debug --disable-rpath --disable-static --with-pic --with-layout=GNU --without-pear --enable-bcmath --enable-calendar --enable-ctype --enable-dba --with-db4=/usr --without-gdbm --with-qdbm=/usr --enable-inifile --enable-flatfile --enable-dom --enable-exif --with-gettext=/usr --enable-fileinfo --enable-filter --enable-ftp --enable-hash --with-iconv --with-pcre-regex=/usr --enable-pdo --enable-mbregex --enable-mbregex-backtrack --enable-mbstring --enable-phar --enable-posix --enable-mysqlnd --enable-mysqlnd-compression-support --with-zlib-dir=/usr --with-openssl=yes --enable-libxml --enable-session --enable-simplexml --enable-soap --enable-sockets --enable-tokenizer --enable-xml --enable-xmlreader --enable-xmlwriter --with-mhash=yes --enable-sysvmsg --enable-sysvsem --enable-sysvshm --enable-zip --with-system-tzdata --enable-dtrace --enable-pdo --enable-mysqlnd --enable-pcntl --with-libedit=shared,/usr
```

and for external modules (as they are built separately):
```
	mkdir -p ext-build
	cd ext-build
	../configure --build=x86_64-linux-gnu --prefix=/usr --includedir=\${prefix}/include --mandir=\${prefix}/share/man --infodir=\${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --libdir=\${prefix}/lib/x86_64-linux-gnu --libexecdir=\${prefix}/lib/x86_64-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --prefix=/usr --enable-cli --disable-cgi --with-config-file-path=/etc/php/7.0/apache2 --with-config-file-scan-dir=/etc/php/7.0/apache2/conf.d --build=x86_64-linux-gnu --host=x86_64-linux-gnu --libdir=\${prefix}/lib/php --libexecdir=\${prefix}/lib/php --datadir=\${prefix}/share/php/7.0 --program-suffix=7.0 --sysconfdir=/etc --localstatedir=/var --mandir=/usr/share/man --disable-all --disable-debug --disable-rpath --disable-static --with-pic --with-layout=GNU --without-pear --enable-bcmath --enable-calendar --enable-ctype --enable-dba --with-db4=/usr --without-gdbm --with-qdbm=/usr --enable-inifile --enable-flatfile --enable-dom --enable-exif --with-gettext=/usr --enable-fileinfo --enable-filter --enable-ftp --enable-hash --with-iconv --with-pcre-regex=/usr --enable-pdo --enable-mbregex --enable-mbregex-backtrack --enable-mbstring --enable-phar --enable-posix --enable-mysqlnd --enable-mysqlnd-compression-support --with-zlib-dir=/usr --with-openssl=yes --enable-libxml --enable-session --enable-simplexml --enable-soap --enable-sockets --enable-tokenizer --enable-xml --enable-xmlreader --enable-xmlwriter --with-mhash=yes --enable-sysvmsg --enable-sysvsem --enable-sysvshm --enable-zip --with-system-tzdata --enable-dtrace --enable-pdo --enable-mysqlnd --with-bz2=shared,/usr --with-curl=shared,/usr --with-enchant=shared,/usr --with-gd=shared,/usr --enable-gd-native-ttf --with-jpeg-dir=shared,/usr --with-xpm-dir=shared,/usr/X11R6 --with-png-dir=shared,/usr --with-freetype-dir=shared,/usr --with-vpx-dir=shared,/usr --with-gmp=shared,/usr --with-imap=shared,/usr --with-kerberos --with-imap-ssl=yes --with-interbase=shared,/usr --with-pdo-firebird=shared,/usr --enable-intl=shared --enable-json=shared --with-ldap=shared,/usr --with-ldap-sasl=/usr --with-mcrypt=shared,/usr --with-mysqli=shared,mysqlnd --with-pdo-mysql=shared,mysqlnd --with-unixODBC=shared,/usr --with-pdo-odbc=shared,unixODBC,/usr --with-pgsql=shared,/usr --with-pdo-pgsql=shared,/usr --with-pspell=shared,/usr --with-libedit=shared,/usr --with-recode=shared,/usr --with-snmp=shared,/usr --with-sqlite3=shared,/usr --with-pdo-sqlite=shared,/usr --with-pdo-dblib=shared,/usr --with-tidy=shared,/usr --with-xmlrpc=shared,/usr --with-xsl=shared,/usr --enable-opcache --enable-opcache-file --enable-huge-code-pages
```

Debian also carry around 46 patches at the moment (although I am trying to reduce the number). One of the patches strips the configure command as it's not really useful in the setup where each SAPI and ext/ modules is built independently.

Anyway - I'll try to update systzdata to v13 from Fedora repository, last time I've seen segfaults like this, it was caused by this common distribution patch. I have just uploaded 7.0.2-3 with updated patch, let me know whether this helped or not.

**Comment by sisve** [comment]
I presume you mean 7.0.3-2 (and not 7.0.2-3). I'll try it out as soon as it's ready; I guesstimate that it will be about an hour further of compilations to be done and will report back once I've tried the update.

**Comment by oerdnj** [comment]
Yes, I meant `7.0.3-2`, and amd64 trusty build just finished: https://launchpad.net/~ondrej/+archive/ubuntu/php/+build/8973690

**Comment by sisve** [comment]
I'm waiting for `apt-get update; apt-get upgrade` to find any changes, but it seems the package list is cached. I'll just have to try out this thing called "patience"... ;)

**Comment by oerdnj** [comment]
You can download the packages by hand from the build page :), if you are really impatient.

**Comment by sisve** [comment]
I still get a segmentation fault with 7.0.3-2. I'll see if I can do the compilation manually without optimizations to get more debug information out of the stacktraces. https://gist.github.com/sisve/04832d5d123b72c79303

**Comment by sisve** [comment]
I tried to compile it using the parameters from the build log, but apparently I lacked a lot of referenced packages. Anyhow, I managed to get a configure that worked for me, compile, and run my test-suite. It now ends with a "php: /root/php-src/Zend/zend_execute.h:275: zend_vm_stack_free_call_frame_ex: Assertion `(executor_globals.vm_stack_top) > (zval *) (executor_globals.vm_stack) && (executor_globals.vm_stack_end) > (zval *) (executor_globals.vm_stack) && (executor_globals.vm_stack_top) <= (executor_globals.vm_stack_end)' failed.
Aborted (core dumped)". 

The backtrace is identical with https://bugs.php.net/bug.php?id=71474 which was fixed in 6039d2d91454e5dc09996d2988d2f38ad8c316ae, currently only part of the master branch. I am guessing that this will be fixed in the next 7.0.4 release.

I'm closing this issue since I finally managed to reproduce the (or perhaps another) error based on the php-source without the debian specific packages or your PPA.

Thanks for your help and information in debugging this!

**Comment by oerdnj** [comment]
Thanks for debugging this, much appreciated. Should I pull the patch into 7.0.3 or you'll happy to wait? I might want to do that so it gets into Xenial (from Debian unstable).

**Comment by oerdnj** [comment]
Also for future reference: `apt-get build-dep php7.0` will install all required packages.

**Comment by sisve** [comment]
I think you should wait so that your releases matches the releases from the upstream php repo, including the segmentation faults and failing asserts. I have no problem waiting; I am only affected by this issue due to my idea to run my test-suite in php7 on my Jenkins machine. I do not use php7 during development or in production (yet).

**Comment by oerdnj** [comment]
Ok, then I'll just reopen the issue and fix it when 7.0.4 is released.

**Comment by oerdnj** [comment]
`php7.0_7.0.3-5+deb.sury.org~*+1` have just been uploaded with the fix pulled out from git to help Ubuntu get the fixed version in time as the DebianImportFreeze is tomorrow.
